### PR TITLE
Identifiable and Numbers

### PR DESCRIPTION
**Pull request by mshinwell**
Flambda uses this new "Identifiable" module to ensure consistent provision of commonly-used data structures across a wide range of types.

There is also a "Numbers" module that provides this functionality for integers and floating-point numbers.

### CODE REVIEW

**Comment by alainfrisch** [comment]
I appreciate the efforts of extracting from flambda changes which can be upstreamed independently and require changes to parts not directly related to flambda.  But for this specific PR, I don't see the rationale of including new code, unused in the rest of the compiler, independently of flambda.

**Comment by gasche** [comment]
I do see some opportunities for code reuse in other parts of the compiler (just having a quick look, for example `Interf.IntPairSet` could use `Identifiable.Pair(Numbers.Int)(Numbers.Int)`, and `Ast_mapper` reimplements the proposed `Map.of_list`), but I would worried about potential module name conflicts for people using `compiler-libs`; maybe we could at least put the content of `numers.ml` in `Misc`?

**Comment by alainfrisch** [comment]
> could use Identifiable.Pair(Numbers.Int)(Numbers.Int)

(This would potentially degrade performances in a time-critical part.)

> ; maybe we could at least put the content of numers.ml in Misc?

What's the point of including a module in the compiler code base if it is not used anywhere?

**Comment by lpw25** [comment]
> What's the point of including a module in the compiler code base if it is not used anywhere?

Since there won't be a release that merges this and doesn't merge flambda, what does it matter. 

Doing it this way makes it easier to review the code.

**Comment by alainfrisch** [comment]
There won't be enough man power to review all the new code in flambda, so one should perhaps focus on critical parts, and on testing.  If we apply the same level of review per line of code than for typical PRs, the merge will not happen in the next year or so.

**Comment by alainfrisch** [comment]
What does the name "Identifiable" mean?

**Comment by chambart** [comment]
There will be some part of the interfering patch relying on this one. This is proposed separately to avoid polluting those other pull requests.

**Comment by lpw25** [comment]
> There won't be enough man power to review all the new code in flambda, so one should perhaps focus on critical parts, and on testing.

Sorry, I should have been clearer. I think the idea is to make the closure.ml based path reviewable via pull requests.

**Comment by gasche** [comment]
> I think the idea is to make the closure.ml based path reviewable via pull requests.

I think this is a good idea and laudable goal.

**Comment by gasche** [comment]
I merged the PRs that felt mergeable and relatively consensual. The remaining ones could benefit from further improvements or discussions.

It is not clear to me which quality level we should demand for these PRs. I _suspect_ that the flambda people hope for a quicker treatment than others, but in absence of clear guideline/consensus I personally applied my usual standards -- basically that the PR be perfect.

**Comment by mshinwell** [comment]
The main reason why I'm doing separate pull requests is actually to make people aware of what changes are required for flambda that do not lie in the completely new parts of the compiler.

This approach also helps to ensure that we don't accidentally push something to trunk, hidden amongst a very large diff, which was not intended.  It also helps to view some of these "odds and ends" changes in isolation, hopefully improving them further (they tend to be the more unloved parts).

As regards @alainfrisch 's point about code review, it should be noted that the flambda code has received a lot of scrutiny from three people, and continues to receive that.

@alainfrisch : The [Identifiable] name is used to match Core_kernel.  I agree it may not be the best name in the world, but the consistency is useful.

**Comment by Drup** [comment]
(As an external observer, I do appreciate the breakdown in sensible episodes, even if I don't have anything to say on this precise episode. :)

**Comment by mshinwell** [comment]
@gasche I'm going to argue that we should not worry about naming conflicts for compiler-libs users.  For a start there are already potentially problematic names ("Types" for example).  Second, the number of users of this library is probably quite small, and the users are probably for the most part the kind of people who will appreciate why the situation arises (e.g. because it keeps "pack" out of the critical path).  Third, we will hopefully be able to resolve this if a proposal for namespacing gets accepted in due course (and I sense we are getting closer on that front; @lpw25 said for example that his thing about explicitly marking .cmi files as opaque would help).  Fourth, flambda has various standard names that I would like not to rename (such as "Variable").

**Comment by mshinwell** [comment]
( @lpw25 points out that this "Numbers" thing might be more of a problem because it's in the common parts, rather than the flambda names, which we can hopefully confine to the "opt" library.  I'm still unsure myself that there is any problem, though.)

**Comment by gasche** [comment]
Given how easy it is to move `Numbers` to a submodule of `Misc`, I see no reason _not_ to do it. For Identifiable it would be a bit less palatable so maybe we can leave it as it is.

**Comment by mshinwell** [comment]
I dislike "Misc" because it's a cop-out on naming.  It's like "loop", "data", etc.  However, if this is the consensus, I will change it.

**Comment by mshinwell** [comment]
We can't move Numbers into Misc without moving Identifiable too, which has previously been said to be undesirable.  So I think we're going to merge this as-is.

**Comment by damiendoligez** [comment]
OK for merging.

**Comment by garrigue** [comment]
When I tried to compile the manual, I got an ocamldoc error with numbers.mli.
Could you fix it.
Seems to be related with using curly braces inside an ocamldoc comment.

File: utils/numbers.ml [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+(*   All rights reserved.  This file is distributed under the terms of    *)
+(*   the GNU Lesser General Public License version 2.1, with the          *)
+(*   special exception on linking described in the file ../LICENSE.       *)
+(*                                                                        *)
+(**************************************************************************)
+
+module Int_base = Identifiable.Make (struct
+  type t = int
+
+  let compare x y = x - y
```
Comment (oandrieu) [inline utils/numbers.ml:20]: Shouldn't that be `Pervasives.compare` instead ? `x - y` can overflow …

**Pull request merged**